### PR TITLE
Jay - PERMISSION: Toggle Tangible Permission

### PR DESCRIPTION
**Pull request by Alforoan**
# Description
This PR enables the user with the Edit Time Entry Toggle Tangible permission to be able to toggle the checkbox of any user's tangible on time entry.

## Related PRS (if any):
This frontend PR is related to the [#713](https://github.com/OneCommunityGlobal/HGNRest/pull/713) backend PR.

…

## Main changes explained:
- Added a new permission label for Edit Time Entry Toggle Tangible permission.
- Added conditional logic so that the user with the permission can check or uncheck the tangible on time entry.

## How to test (please watch the video):
1. check into current branch
2. do `npm install` and `npm run start:local` to run this PR locally
3. Clear site data/cache
4. log in with an owner account
5. go to Other Links -> Permissions Management -> Manage User Permissions -> Choose a user without **Toggle Tangible Time Others** permission under Timelog Management, Edit Time Entries -> click on Add  -> Scroll down then submit
6. verify that the user who was given the permission is now able to check and uncheck the box next to Tangible: on another user's dashboard for time log entries.

## Screenshots or videos of changes:

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/118868270/9633a908-3871-4ab6-8584-437650899406

## Note:
**Unmute the video**

### CODE REVIEW

File: src/components/Timelog/TimeEntry.jsx [hunk 1]
```diff
@@ -84,6 +84,20 @@ const TimeEntry = (props) => {
     //permission to delete any time entry on their own time logs tab
     dispatch(hasPermission('deleteTimeEntry')) ||
     //default permission: delete own sameday tangible entry
+    (!isTangible && isAuthUser && isSameDay);
+
+  const canEdit =
```
Comment (Nahiyan-16) [inline src/components/Timelog/TimeEntry.jsx:89]: This variable name was already declared.
Maybe attempt to comment out previous declaration.

**Review by Nahiyan-16 — CHANGES_REQUESTED** [review]
## Syntax Error
Hi Jay, there seems to be a syntax error on the time entry file.
I did try testing by commenting out one of the double declarations but was unable to produce results.  
Screenshot 2024-01-24 084401

**Review by Renanluizssx — CHANGES_REQUESTED** [review]
Hi @Alforoan, i tested your application and i got to add an intangible time entry, but i didn't get to check tangible, look my review below please:

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/91035340/f776f6a4-4a9e-4ff9-bc72-065648a01196

**Comment by Alforoan** [comment]
@Nahiyan-16 @Renanluizssx Hi, could you please test this PR again?

**Review by Nahiyan-16 — COMMENTED** [review]
## Toggle not working for me
Hi Jay, I can't seem to toggle off "Tangible Time Others". I was trying to see what the toggled-off version would look like so I made sure that this option was off and submitted it. I logged onto the admin account where I made the change and viewed another person's dashboard where I could toggle their tangible time even though it was off. Here's a video of it: 
 

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/71025942/970bc51c-b01f-4ffa-97ad-d8817c173711

Please let me know if I was doing anything wrong.

**Comment by galgalinidhi** [comment]
Hi @Alforoan, I tested the PR with a Volunteer account and it works as expected.


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/44816413/c4982b24-c086-403d-9a86-e132b8673532

However, when I test the functionality for an Admin account, I can still toggle Tangible Time Others even though the permission isn't added.


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/44816413/b6445426-8bdf-4cf0-b67a-af868c33e6f3

**Review by Pandani07 — CHANGES_REQUESTED** [review]
I have tested this PR and the functionality works as expected. But I can also toggle tangible on or off with my admin account as well. I never added this permission to my admin account


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/44672833/33e524a4-c689-49ac-a467-d28aaf20b51f


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/44672833/dff7a579-d537-41b0-9d4d-81cf3e450362

File: src/components/Timelog/TimeEntry.jsx [hunk 2]
```diff
@@ -163,15 +170,15 @@ const TimeEntry = (props) => {
             </p>
             <div className='mb-3'>
             {
-              canEdit 
+              (canEdit || canEditTimeEntryToggleTangible)
```
Comment (nathanah) [inline src/components/Timelog/TimeEntry.jsx:173]:
```suggestion
              (canEditTimeEntryToggleTangible)
```

File: src/components/Timelog/TimeEntry.jsx [hunk 3]
```diff
@@ -163,15 +170,15 @@ const TimeEntry = (props) => {
...
                 ? ( 
                     <>
                       <span className="text-muted">Tangible:&nbsp;</span>
                       <input
                           type="checkbox"
                           name="isTangible"
                           checked={isTangible}
-                          disabled={!canEdit}
+                          disabled={!(canEdit || canEditTimeEntryToggleTangible)}
```
Comment (nathanah) [inline src/components/Timelog/TimeEntry.jsx:181]:
```suggestion
                          disabled={!canEditTimeEntryToggleTangible}
```

**Review by nathanah — CHANGES_REQUESTED** [review]
`editTimeEntryToggleTangible` should be replacing `editTimeEntry`

**Review by XiaohanMeng — CHANGES_REQUESTED** [review]
Hi, I tested this PR using the owner, volunteer, and admin accounts. After the owner granted permissions to the volunteer account, it functioned very well. However, I encountered the same issue as the preview reviewer. Even without granting permissions to the admin account, it was still able to toggle Tangible Permission. And some error jumped out at my backend terminal. Please see the reference video and screenshot. 
PR1862+713-error

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/113215317/104f500d-6d36-4091-b03e-18ad845de1a8

**Review by Jingyii800 — CHANGES_REQUESTED** [review]
Hi @Alforoan, I have tested this feature with backend PR. It mostly works, but there is one issue with the correspondence. When I have two time logs 1 and 2, I **clicked on "tangible" for time log 2**, but actually **time log 1 became tangible**. Same situation when I tried to make two time logs untangible. Whatever I clicked on,  time log 1 is always the first to be changed.  Please check more details in the video. 

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/112589476/407a3f9d-5df9-4dc5-b5ad-df51b0e0ff66

**Review by ClemarNunes — CHANGES_REQUESTED** [review]
I tested the PR with a Volunteer account and it worked as expected. However, when testing the functionality with an administrator account, I found that I could still toggle "Tangible Time Others" even though the permission was not added.


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/134466562/d19b2c06-d38c-4685-b741-88a3228c1393

**Review by Gugor99 — CHANGES_REQUESTED** [review]
### Report
Hey Jay, 
I tried your implementation but it wouldn't work properly on my machine. I firstly logged as owner to give permission to my volunteer account to toggle tangible time others. I then logged out and back in as volunteer, tried to go to my own testrole account and uncheck the tangible box of the time entry just did. I did not have the ability to change the value of the checkbox. I also added intangible time on that account, always using the volunteer's and still wasn't able to change it. Finally logged in as owner with the same result, I wasn't able to change the value of the checkbox with any account, maybe I did something wrong. Also a message logged in the console, see screenshots for reference.
### Screenshots
Cattura1
Cattura2
Cattura3
Cattura5
Cattura6

**Review by brandta-1 — CHANGES_REQUESTED** [review]
# Review
Successfully changed a volunteer user's permissions and toggled tangibility of another user's time entries on the other user's dashboard. 
## Error
Tangible toggle checkbox in ```src/components/Timelog/TimeEntry.jsx``` must be clicked twice for the tangibility to toggle. No console errors were being logged in the browser.

# Video

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/116298512/9f630f3d-be5e-45ff-8bb9-0368d51bb3ac


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/116298512/51687b21-34b4-4aae-9bba-a4b75c37f048

**Review by Weiyao-Li — APPROVED** [review]
Hi, I have tested it. Worked as expected. Great work
image
image
image

**Comment by DiegoSalas27** [comment]
Hi Jay,

I have tested your functionality and it works as described in your PR. Good work.


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/26840798/42b82a37-13d0-45e6-9174-3a5b51baf0ee

**Review by Sarthak32 — COMMENTED** [review]
I tested this it works as expected  but what I found despite not giving permissions to  account, it still had the ability to toggle Tangible Permission.

Screenshot 2024-03-10 122119
Screenshot 2024-03-10 122138
Screenshot 2024-03-10 125017

**Review by psharma1984 — APPROVED** [review]
Hi @Alforoan The PR is working as expected

[screen-capture (1).webm](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/123996264/cd21cc04-26fe-406a-a0c8-30d9cfaccf3f)

**Comment by rajnada** [comment]
Hi,
After following all the steps and reviewing the PR, when I log in using my volunteer account, there is no toggle button outside the description box. When I click on the edit button, only a new dialogue box opens, and in that, I can see the option to add/edit Tangible time using the toggle button.

Here below I can show you the images
Assign permission:
1862B

Here I am not able to see the toggle button:
1862A

**Review by iven-yao — APPROVED** [review]
following screenshots show that before and after granted permission `Toggle Tangible Time Others` on a user. User could see the checkbox on other's dashboard and switch it to tangible or intangible as expected. However, this permission will only apply to the user after re-login to the system.

image
image
image

**Review by hetvi11110 — COMMENTED** [review]
As mentioned, I have reviewed the PR, and the Toggle Tangible Time Others is working fine. However, the Toggle Tangible Time Self  is not functioning as expected. Please take a look at the reference video for more information.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/13557067/2493d0d1-7051-43fc-b9d1-3f7ca83fa24a

**Review by SRamen1999 — COMMENTED** [review]
After checking this PR I encountered some bugs. When I logged in using an owner account I would keep getting a blank white screen on the dashboard section. I even cleared the local/session storage and cookies.

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/56932664/78f944a3-d1f1-4fa5-bd40-2bd4f91542e6

However when given the toggle permission it works as expected

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/56932664/c20437f8-7289-4a54-9e0a-b1c4c90c5f19

**Review by Tiantian-C — COMMENTED** [review]
I have tested this PR.
I first log in as an owner and change the permission of an admin user and then this admin user can toggle tangible permission.


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/122494725/271209ac-a12a-46b7-b5da-ae74c5b07d66

**Review by Tiantian-C — COMMENTED** [review]
I have tested this PR. 
I first logged in as an owner and changed the permission of an admin,then I logged in as this admin user. I found that I can toggle tangible permission.


https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/122494725/94b348cf-0332-4f77-8a5c-b880d9728e92

**Review by kylepham — APPROVED** [review]
PR worked like described. Checkbox didn't show up without permission to edit, and it did when I granted the permission.

Screenshot 2024-03-31 at 4 20 52 PM

https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/35129229/37407629-1a03-4fd3-bda3-144a5d76a70f

**Pull request closed without merging**